jmdaemon/jmdaemon/onionmc.py

```diff
@@ -0,0 +1,1210 @@
+
+    def message(self, message: OnionCustomMessage) -> None:
+        self.sendLine(message.encode())
+
+class OnionLineProtocolFactory(protocol.ServerFactory):
+    """ This factory allows us to start up instances
+    of the LineReceiver protocol that are instantiated
+    towards us.
+    """
+    protocol = OnionLineProtocol
+
+    def __init__(self, client: 'OnionMessageChannel'):
+        self.client = client
+        self.peers = {}
+
+    def register_connection(self, p: OnionLineProtocol) -> None:
+        # make a local control message registering
+        # the new connection
+        peer_location = network_addr_to_string(p.transport.getPeer())
+        self.client.register_connection(peer_location, direction=0)
+        self.peers[peer_location] = p
+
+    def register_disconnection(self, p: OnionLineProtocol) -> None:
+        # make a local control message registering
+        # the new connection
+        peer_location = network_addr_to_string(p.transport.getPeer())
+        self.client.register_disconnection(peer_location)
+        if not peer_location in self.peers:
+            log.warn("Disconnection event registered for non-existent peer.")
+            return
+        del self.peers[peer_location]
+
+    def receive_message(self, message: OnionCustomMessage,
+                        p: OnionLineProtocol) -> None:
+        self.client.receive_msg(message, network_addr_to_string(
+            p.transport.getPeer()))
+
+    def send(self, message: OnionCustomMessage, destination: str) -> bool:
+        if not (destination in self.peers):
+            log.warn("sending message {}, destination {} was not in peers {}".format(
+                message.encode(), destination, self.peers))
+            return False
+        proto = self.peers[destination]
+        proto.message(message)
+        return True
+
+class OnionClientFactory(protocol.ClientFactory):
+    """ We define a distinct protocol factory for outbound connections.
+    Notably, this factory supports only *one* protocol instance at a time.
+    """
+    protocol = OnionLineProtocol
+
+    def __init__(self, message_receive_callback: Callable,
+                 connection_callback: Callable,
+                 disconnection_callback: Callable):
+        self.proto_client = None
+        # callback takes OnionCustomMessage as arg and returns None
+        self.message_receive_callback = message_receive_callback
+        # connection callback, no args, returns None
+        self.connection_callback = connection_callback
+        # disconnection the same
+        self.disconnection_callback = disconnection_callback
+
+    def register_connection(self, p: OnionLineProtocol) -> None:
+        self.proto_client = p
+        self.connection_callback()
+
+    def register_disconnection(self, p: OnionLineProtocol) -> None:
+        self.proto_client = None
+        self.disconnection_callback()
+
+    def send(self, msg: OnionCustomMessage) -> bool:

```

Comment:
Yes, thanks, you picked up on something tricky here.
There are two ways we send out a message: either via the `OnionClientFactory` object, if we have an outbound connection to the peer; in this case we call the `sendLine` method of the underlying protocol instance directly (one specific protocol instance for that specific peer) and it returns `None` (the actual sending of the line is done asynchronously; it could fail of course but higher layers of the stack just deal with that).
The other way is when we are serving an onion and we have an inbound connection to use; then we send out via the `OnionLineProtocolFactory` object, which *checks to see if the peer is available* (because one of these `Factory` objects is supporting multiple such inbound protocol instances, so it's keeping an internal list of who it currently has available); if the peer is not currently present, it returns `False` as in "we couldn't even try to send that message; peer is not here", otherwise it returns `True` as in "we succeeded in at least trying to send it".

Reply:
So I will ensure that both versions return `bool`; it's slightly weird in that one will *always* return `True` for now (unless I missed something), but it will give logical consistency. Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1182/commits/f74b04352c06c6293976ae173964b34672ac1390